common/src/main/java/com/microsoft/identity/common/internal/platform/SP800108KeyGen.java

```diff
@@ -0,0 +1,119 @@
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import lombok.NonNull;
+
+/**
+ * This class is a key derivation function for SP800 108 key generation.
+ */
+public class SP800108KeyGen {
+
+    public static final byte[] BIG_ENDIAN_INT_256 = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(256).array();
+
+    /**]
+     * Generate a derived key given a starting key.
+     * @param key the basis for the key material.
+     * @param label a label for the key.
+     * @param ctx the key context.
+     * @return a derived key.
+     * @throws IOException
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    public static byte[] generateDerivedKey(@NonNull final byte[] key, @NonNull final byte[] label, @NonNull final byte[] ctx)
+            throws IOException, InvalidKeyException, NoSuchAlgorithmException {
+        if (ctx == null) {
+            return null;
+        }
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        stream.write(label);
+        stream.write(0x0);
+        stream.write(ctx);
+
+        stream.write(BIG_ENDIAN_INT_256);
+
+        byte[] pbDerivedKey = constructNewKey(key, stream.toByteArray());
+        return Arrays.copyOf(pbDerivedKey, 32);
+    }
+
+    private static byte[] constructNewKey(byte[] keyDerivationKey, byte[] fixedInput)
+            throws IOException, NoSuchAlgorithmException, InvalidKeyException {
+        byte ctr;
+        byte[] cHMAC;
+        byte[] keyDerivated;
+        byte[] dataInput;
+
+        int len;
+        int numCurrentElements;
+        int numCurrentElementsBytes;
+        int outputSizeBit = 256;
+
+        numCurrentElements = 0;
+        ctr = 1;
+        keyDerivated = new byte[outputSizeBit / 8];
+        final SecretKeySpec keySpec = new SecretKeySpec(keyDerivationKey, "HmacSHA256");
+        final Mac hmacSHA256 = Mac.getInstance("HmacSHA256");
+
+        do {
+            dataInput = updateDataInput(ctr, fixedInput);
+            hmacSHA256.reset();
+            hmacSHA256.init(keySpec);
+            hmacSHA256.update(dataInput);
+            cHMAC = hmacSHA256.doFinal();
+            if (256 >= outputSizeBit) {
+                len = outputSizeBit;
+            } else {
+                len = Math.min(256, outputSizeBit - numCurrentElements);
+            }
+
+            numCurrentElementsBytes = numCurrentElements / 8;
+            System.arraycopy(cHMAC, 0, keyDerivated, numCurrentElementsBytes, 32);
+            numCurrentElements = numCurrentElements + len;
+            ctr++;
+        } while (numCurrentElements < outputSizeBit);
+        return keyDerivated;
+    }
+
+    private static byte[] updateDataInput(final byte ctr, @NonNull final byte[] fixedInput) throws IOException {
+        final ByteArrayOutputStream tmpFixedInput = new ByteArrayOutputStream(fixedInput.length + 4);
```

Comment:
I don't claim to know about the bitshifts, or extra alloc of 4 bytes. I also don't know the ins/outs of KD, so maybe this is just over my head. I think this code was ported from CPP? I just hope I don't need to one day debug it 😨

Reply:
Yeah, this was ported from C++ which ported it from objc. It lets us avoid a dependency on BouncyCastle, so far.